### PR TITLE
Prevent CachedField and CachedMethod from leaking access permissions …

### PR DESCRIPTION
**Pull request by dpolivaev**
…to scripts

https://issues.apache.org/jira/browse/GROOVY-8163

### CODE REVIEW

File: src/main/org/codehaus/groovy/reflection/AccessPermissionChecker.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+package org.codehaus.groovy.reflection;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.lang.reflect.ReflectPermission;
+
+import groovy.lang.GroovyObject;
+
+class AccessPermissionChecker {
```
Comment (dpolivaev) [inline src/main/org/codehaus/groovy/reflection/AccessPermissionChecker.java]: done

File: src/main/org/codehaus/groovy/reflection/AccessPermissionChecker.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+	private static final ReflectPermission REFLECT_PERMISSION = new ReflectPermission("suppressAccessChecks");
+
+	static private void checkAccessPermission(Class<?> declaringClass, final int modifiers, boolean isAccessible) {
+		final SecurityManager securityManager = System.getSecurityManager();
```
Comment (dpolivaev) [inline src/main/org/codehaus/groovy/reflection/AccessPermissionChecker.java]: Signature changed. The check is performed only if (securityManager != null && isAccessible)

File: src/main/org/codehaus/groovy/reflection/AccessPermissionChecker.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+					securityManager.checkCreateClassLoader();
+				}
+		}
+	}
+
+	private static boolean packageCanNotBeAddedAnotherClass(Class<?> declaringClass) {
+		return declaringClass.getName().startsWith("java.");
+	}
+
+	static public void checkAccessPermission(Method method) {
```
Comment (dpolivaev) [inline src/main/org/codehaus/groovy/reflection/AccessPermissionChecker.java]: fixed

File: src/main/org/codehaus/groovy/reflection/AccessPermissionChecker.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+		checkAccessPermission(method.getDeclaringClass(), method.getModifiers(), method.isAccessible()
+		);
+	}
+
+	public static void checkAccessPermission(Field field) {
```
Comment (dpolivaev) [inline src/main/org/codehaus/groovy/reflection/AccessPermissionChecker.java]: fixed

File: src/main/org/codehaus/groovy/reflection/CachedField.java [hunk 5]
```diff
@@ -51,6 +52,12 @@ public int getModifiers() {
      */
     public Object getProperty(final Object object) {
         try {
+            AccessPermissionChecker.checkAccessPermission(field);
+        }
+        catch (AccessControlException ex) {
+            throw new IllegalArgumentException("Illegal access to field" + " " + field.getName());
```
Comment (dpolivaev) [inline src/main/org/codehaus/groovy/reflection/CachedField.java]: To use GroovyRuntimeException we need to patch also MetaClassImpl.getProperty . It was not possible with byte buddy, but as a groovy patch it is possible. I shall do it in a separate commit.

File: src/main/org/codehaus/groovy/reflection/CachedField.java [hunk 6]
```diff
@@ -65,6 +72,12 @@ public Object getProperty(final Object object) {
      * @throws RuntimeException if the property could not be set
      */
     public void setProperty(final Object object, Object newValue) {
+        try {
+            AccessPermissionChecker.checkAccessPermission(field);
+        }
+        catch (AccessControlException ex) {
+            throw new IllegalArgumentException("Illegal access to field" + " " + field.getName());
```
Comment (dpolivaev) [inline src/main/org/codehaus/groovy/reflection/CachedField.java]: As above

File: src/main/org/codehaus/groovy/reflection/CachedMethod.java [hunk 7]
```diff
@@ -90,6 +91,12 @@ public CachedClass getDeclaringClass() {
 
     public final Object invoke(Object object, Object[] arguments) {
         try {
+            AccessPermissionChecker.checkAccessPermission(cachedMethod);
+        }
+        catch (AccessControlException ex) {
+            throw new InvokerInvocationException(new IllegalArgumentException("Illegal access to method" + cachedMethod.getName()));
```
Comment (dpolivaev) [inline src/main/org/codehaus/groovy/reflection/CachedMethod.java]: As above

File: src/main/org/codehaus/groovy/reflection/CachedMethod.java [hunk 8]
```diff
@@ -124,6 +131,12 @@ public String getSignature() {
     }
 
     public final Method setAccessible() {
+        try {
+            AccessPermissionChecker.checkAccessPermission(cachedMethod);
+        }
+        catch (AccessControlException ex) {
+            throw new IllegalArgumentException("Illegal access to method" + cachedMethod.getName());
```
Comment (dpolivaev) [inline src/main/org/codehaus/groovy/reflection/CachedMethod.java]: As above

File: src/main/org/codehaus/groovy/reflection/CachedMethod.java [hunk 9]
```diff
@@ -324,6 +337,12 @@ else if (o2 instanceof CachedMethod)
     }
 
     public Method getCachedMethod() {
+        try {
+            AccessPermissionChecker.checkAccessPermission(cachedMethod);
+        }
+        catch (AccessControlException ex) {
+            throw new IllegalArgumentException("Illegal access to method" + cachedMethod.getName());
```
Comment (dpolivaev) [inline src/main/org/codehaus/groovy/reflection/CachedMethod.java]: As above

**Comment by dpolivaev** [comment]
Changes suggested in the review are implemented.
See also https://issues.apache.org/jira/browse/GROOVY-8163?focusedCommentId=16009695&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16009695

File: src/main/groovy/lang/MetaClassImpl.java [hunk 10]
```diff
@@ -1832,6 +1832,9 @@ public Object getProperty(Class sender, Object object, String name, boolean useS
             } catch (IllegalArgumentException e) {
                 // can't access the field directly but there may be a getter
                 mp = null;
+            } catch (GroovyRuntimeException e) {
+                // can't access the field directly but there may be a getter
+                mp = null;
```
Comment (jwagenleitner) [inline src/main/groovy/lang/MetaClassImpl.java]: I don't understand why this was added.  Was there a test that failed because of the other changes that required this.  I would have assumed an `AccessControlException` from `CachedField.getProperty` would be treated similar to the `IllegalAccessException` in the same method.

Comment (dpolivaev) [inline src/main/groovy/lang/MetaClassImpl.java]: I do not have unit test to explain this catch block. I do have the integration test https://issues.apache.org/jira/browse/GROOVY-8163?focusedCommentId=16009695&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16009695. The problem is that for some classes C with private member called "name" class property C.class.name which corresponds to java calls C.class.getName() does not work unless this catch block is added.

File: src/main/groovy/lang/MetaClassImpl.java [hunk 11]
```diff
@@ -1832,6 +1832,9 @@ public Object getProperty(Class sender, Object object, String name, boolean useS
             } catch (IllegalArgumentException e) {
                 // can't access the field directly but there may be a getter
                 mp = null;
+            } catch (GroovyRuntimeException e) {
```
Comment (blackdrag) [inline src/main/groovy/lang/MetaClassImpl.java]: At this point mp should be a CachedField. Since AccessPermissionChecker.checkAccessPermission(field); may now throw a GroovyRuntimeException, we have here the requirement to catch it. Problem is, that mp might not be a CachedField. If we will call a user method here instead, this would lead to a wrong flow. Thus I think the GroovyRuntimeException usage is a bad idea. Better make a new Exception, that extends GroovyRuntimeException and throw that in AccessPermissionChecker, which we then can catch

File: src/main/org/codehaus/groovy/reflection/CachedMethod.java [hunk 12]
```diff
@@ -90,6 +92,12 @@ public CachedClass getDeclaringClass() {
 
     public final Object invoke(Object object, Object[] arguments) {
         try {
+            AccessPermissionChecker.checkAccessPermission(cachedMethod);
+        }
+        catch (AccessControlException ex) {
+            throw new InvokerInvocationException(new GroovyRuntimeException("Illegal access to method" + cachedMethod.getName()));
+        }
+        try {
             return cachedMethod.invoke(object, arguments);
```
Comment (blackdrag) [inline src/main/org/codehaus/groovy/reflection/CachedMethod.java]: can we move thos try-catch into checkAccessPermission ?

File: src/main/groovy/lang/MetaClassImpl.java (hunk 11 above)
Comment (dpolivaev) [inline src/main/groovy/lang/MetaClassImpl.java]: I think AccessPermissionChecker should only throw exceptions extending from AccessControlException. because it specifically checks access permissions. 

So I suggest to introduce `public class CacheAccessControlException extends AccessControlException`

File: src/main/org/codehaus/groovy/reflection/CachedMethod.java (hunk 12 above)
Comment (dpolivaev) [inline src/main/org/codehaus/groovy/reflection/CachedMethod.java]: I think that InvokerInvocationException is specific to call of invoke() and it should be not thrown  by AccessPermissionChecker directly.

File: src/main/groovy/lang/MetaClassImpl.java (hunk 11 above)
Comment (dpolivaev) [inline src/main/groovy/lang/MetaClassImpl.java]: On the other side if it does not work because AccessControlExceptions are not handled by the calling code properly `public class CacheAccessControlException extends GroovyRuntimeException` could also be taken as a hack.

Comment (dpolivaev) [inline src/main/groovy/lang/MetaClassImpl.java]: How do you see it?

Comment (dpolivaev) [inline src/main/groovy/lang/MetaClassImpl.java]: Done

**Comment by dpolivaev** [comment]
Could you please review the changes I implemented on your behalf?

**Comment by blackdrag** [comment]
+1

**Comment by jwagenleitner** [comment]
@dpolivaev can you squash up the commits and change the tabs in `AccessPermissionChecker` to spaces? If not, just let me know and I can take care of it.  In the future, it's helpful to create a separate branch for the PR to make it easier to rebase and avoid merge commits.

@blackdrag I would guess this is a candidate for 2_5_X and above, but maybe too much of a change for 2.4.12, though maybe not `breaking` in a code sense it could affect those running with security managers.  Do you agree?

**Comment by blackdrag** [comment]
agreed

**Comment by dpolivaev** [comment]
Done

**Comment by jwagenleitner** [comment]
Thanks.
